OST2.design/tools/perl-5.8.0/man/man3/BigIntSupport.3
.\" Automatically generated by Pod::Man v1.34, Pod::Parser v1.13
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "BigIntSupport 3"
.TH BigIntSupport 3 "2002-07-24" "perl v5.8.0" "User Contributed Perl Documentation"
.SH "NAME"
BigIntSupport \- Perl extension for interfacing between Math::BigInt values and text strings.
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 2
\&  use BigIntSupport;
\&  use Math::BigInt;
.Ve
.PP
.Vb 10
\&  # Create Math::BigInt from an arbitrary-sized hex number
\&  my $bigint1 = hex2bigint("0xdeadbeefcafebabe01234567");
\&  my $bigint2 = hex2bigint("0xffffffff");
\&  # Do math
\&  $bigint2 <<= 32;
\&  $bigint1 &= $bigint2;
\&  # Get hex string back
\&  my $string = bigint2hex($bigint1);
\&  # Print it
\&  print "Result is 0x$string\en";
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
This provides some interface functions for the Math::BigInt library,
which is part of the perl standard libary.  Basically, Math::BigInt is
a module for arbitrary-sized integers.  It overloads all the standard
math operators for use with Math::BigInt objects, but it does not
overload either the \fIhex()\fR function or sprintf \*(L"%x\*(R", so it is difficult
to use Math::BigInt objects for large hex numbers.  This module is a
solution to that problem.
.Sh "\s-1EXPORT\s0"
.IX Subsection "EXPORT"
The following functions are exported:
.IP "hex2bigint($hex_string)" 4
.IX Item "hex2bigint($hex_string)"
Interprets \f(CW$hex_string\fR as a hex number (whether or not a leading 0x is
present).  Returns a Math::BigInt object representing the number.  To
convert decimal numbers to Math::BigInt objects, use
Math::BigInt\->new($decimal_string).
.IP "bigint2hex($bigint)" 4
.IX Item "bigint2hex($bigint)"
Formats the given Math::BigInt object as a hexidecimal string and
returns it.  The string has no leading zeros or 0x.
.ie n .IP "bigint_sprintf($bigint, $format)" 4
.el .IP "bigint_sprintf($bigint, \f(CW$format\fR)" 4
.IX Item "bigint_sprintf($bigint, $format)"
Takes a Math::BigInt and a printf-style format string and returns a
string with the appropriately-formatted string.  The format string can
contain only \f(CW%x\fR or \f(CW%d\fR, and may optionally include a width
specification, which may include a leading zero.  In other words,
\&\*(L"%016x\*(R" is legal but \*(L"The answer is \f(CW%d\fR\*(R" is not.  A leading \*(L"%\*(R" is
optional.
.SH "AUTHOR"
.IX Header "AUTHOR"
Jeff Gibson, jgibson@afara.com
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIMath::BigInt\fR\|(3), \fIperl\fR\|(1).


OST2.design/tools/perlmod/BigIntSupport.pm
# ========== Copyright Header Begin ==========================================
# 
# OpenSPARC T2 Processor File: BigIntSupport.pm
# Copyright (C) 1995-2007 Sun Microsystems, Inc. All Rights Reserved
# 4150 Network Circle, Santa Clara, California 95054, U.S.A.
#
# * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER. 
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; version 2 of the License.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#
# For the avoidance of doubt, and except that if any non-GPL license 
# choice is available it will apply instead, Sun elects to use only 
# the General Public License version 2 (GPLv2) at this time for any 
# software where a choice of GPL license versions is made 
# available with the language indicating that GPLv2 or any later version 
# may be used, or where a choice of which version of the GPL is applied is 
# otherwise unspecified. 
#
# Please contact Sun Microsystems, Inc., 4150 Network Circle, Santa Clara, 
# CA 95054 USA or visit www.sun.com if you need additional information or 
# have any questions. 
# 
# ========== Copyright Header End ============================================
package BigIntSupport;

require 5.005_62;
use strict;
use warnings;

require Exporter;
use Math::BigInt;
use Carp;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use BigIntSupport ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
		 hex2bigint
		 bigint2hex
		 bigint_sprintf
		);
our $VERSION = '1.01';


# Preloaded methods go here.

sub hex2bigint {
  my $hexstring = shift;
  $hexstring = 0 unless defined $hexstring;
  $hexstring =~ s/^\s+//;
  $hexstring =~ s/\s+$//;
  $hexstring =~ s/^0[xX]//;

  my $bignum = Math::BigInt->new(0);
  my $shift = 0;

  while(length $hexstring) {
    my $tail = substr $hexstring, -8, 8, "";
    my $tailval = hex($tail);
    $bignum += Math::BigInt->new($tailval) << $shift;
    $shift += 32;
  }

  return $bignum;
}


sub bigint2hex {
  my $bignum = shift;

  if(not ref $bignum) {
    return sprintf("%x", $bignum);
  }

  my $string = "";

  while($bignum) {
    my $low = $bignum & 0xffffffff;
    $string = sprintf("%08x", $low) . $string;
    $bignum >>= 32;
  }
  $string =~ s/^0+//;

  $string = '0' if $string eq '';

  return $string;
}


sub bigint_sprintf {
  my $bigint = shift;
  my $format = shift;

  if($format =~ /^\s*\%?(0)?(\d+)?[lL]*([xd])\s*/) {
    my ($lead, $width, $base) = ($1, $2, $3);

    $width = 0 unless defined $width;
    $base = '' unless defined $base;

    my $string;


    if($base eq 'd') {
      $string = "$bigint";
      $string =~ s/^\+//;
    } else {
      $string = bigint2hex($bigint);
    }

    my $strlen = length($string);
    if($width > $strlen) {
      my $diff = $width - $strlen;
      my $char = defined($lead)? '0' : ' ';
      my $leading = $char x $diff;
      $string = "$leading$string";
    }
    return $string;
  } else {
    croak "Bad format string: $format";
  }
}


1;
__END__
# Below is stub documentation for your module. You better edit it!

=head1 NAME

BigIntSupport - Perl extension for interfacing between Math::BigInt values and text strings.

=head1 SYNOPSIS

  use BigIntSupport;
  use Math::BigInt;

  # Create Math::BigInt from an arbitrary-sized hex number
  my $bigint1 = hex2bigint("0xdeadbeefcafebabe01234567");
  my $bigint2 = hex2bigint("0xffffffff");
  # Do math
  $bigint2 <<= 32;
  $bigint1 &= $bigint2;
  # Get hex string back
  my $string = bigint2hex($bigint1);
  # Print it
  print "Result is 0x$string\n";


=head1 DESCRIPTION

This provides some interface functions for the Math::BigInt library,
which is part of the perl standard libary.  Basically, Math::BigInt is
a module for arbitrary-sized integers.  It overloads all the standard
math operators for use with Math::BigInt objects, but it does not
overload either the hex() function or sprintf "%x", so it is difficult
to use Math::BigInt objects for large hex numbers.  This module is a
solution to that problem.

=head2 EXPORT

The following functions are exported:

=over 4

=item hex2bigint($hex_string)

Interprets $hex_string as a hex number (whether or not a leading 0x is
present).  Returns a Math::BigInt object representing the number.  To
convert decimal numbers to Math::BigInt objects, use
Math::BigInt->new($decimal_string).

=item bigint2hex($bigint)

Formats the given Math::BigInt object as a hexidecimal string and
returns it.  The string has no leading zeros or 0x.

=item bigint_sprintf($bigint, $format)

Takes a Math::BigInt and a printf-style format string and returns a
string with the appropriately-formatted string.  The format string can
contain only %x or %d, and may optionally include a width
specification, which may include a leading zero.  In other words,
"%016x" is legal but "The answer is %d" is not.  A leading "%" is
optional.

=back



=head1 SEE ALSO

Math::BigInt(3), perl(1).

=cut
